File: Soluções/Pairing Parentheses/Java/erickson_stack.java
import java.util.Scanner;
import java.util.Stack;

public class Solution {

    public static char getOpposite(char c) {
        switch (c) {
            case '{':
                return '}';
            case '[':
                return ']';
            case '(':
                return ')';
            case '<':
                return '>';
            case '}':
                return '{';
            case ']':
                return '[';
            case ')':
                return '(';
            case '>':
                return '<';
            default:
                return ' ';
        }
    }

    public static boolean isOpening(char c) {
        return "{[(<".contains("" + c);
    }

    public static boolean check(String str) {
        boolean valid = true;
        Stack<Character> tokenStack = new Stack<Character>();

        for (int i = 0; i != str.length(); i++) {
            char current = str.charAt(i);
            char opposite = getOpposite(current);

            if (opposite != ' ') {
                if (isOpening(current)) {
                    tokenStack.push(current);
                } else {
                    char top = ' ';
                    if (!tokenStack.isEmpty()) {
                        top = tokenStack.pop();
                    }
                    if (top != opposite) {
                        valid = false;
                        break;
                    }
                }
            }
        }

        valid = valid && tokenStack.size() == 0;

        return valid;
    }

    public static void main(String[] args) {
        int testCase, i;
        String str;
        int maxTestCases = 10;
        Scanner scanner = new Scanner(System.in);
        for (i = 0; i != maxTestCases; i++) {
            testCase = scanner.nextInt();
            str = scanner.next();
            System.out.println("#" + (i + 1) + " " + (check(str) ? "1" : "0"));
        }
        scanner.close();
    }

}

File: Soluções/Pairing Parentheses/C++/erickson_stack.cpp
#include <iostream>
#include <stack>
#include <string>

const char get_opposite(const char& c)
{
    switch (c) {
        case '{': return '}';
        case '[': return ']';
        case '(': return ')';
        case '<': return '>';
        case '}': return '{';
        case ']': return '[';
        case ')': return '(';
        case '>': return '<';
        default: return ' ';
    }
}

const std::string opening_token = "{[(<";
const std::string closing_token = "}])>";

const bool is_closing_token(const char& c)
{
    return closing_token.find(c) != std::string::npos;
}

const bool is_opening_token(const char& c)
{
    return opening_token.find(c) != std::string::npos;
}

const bool check(const std::string& str)
{
    unsigned int i;
    bool valid = true;
    std::stack<char> token_stack;
    for (i = 0; i != str.size(); ++i)
    {
        const char c_char = str[i];
        const char o_char = get_opposite(c_char);
        if (o_char != ' ')
        {
            if (is_opening_token(c_char))
            {
                token_stack.push(c_char);
            }
            else
            {
                if (token_stack.size() > 0 && o_char == token_stack.top())
                {
                    token_stack.pop();
                }
                else
                {
                    valid = false;
                    break;
                }
            }
        }
    }
    valid = valid && token_stack.size() == 0;
    return valid;
}

int main(int argc, const char** argv)
{
    const unsigned int maximum_test_cases = 10;
    unsigned int test_case, i;
    std::string str;
    for (i = 0; i != maximum_test_cases; ++i)
    {
        std::cin >> test_case;
        std::cin >> str;
        std::cout << "#" << (i + 1) << " " << (check(str) ? "1" : "0") << std::endl;
    }
    return 0;
}
